### ISSUE TITLE
WebApp stores image

### ISSUE THREAD

**Issue by doyu** [Origami-TinyML/tflm_hello_world#50]
ref: https://github.com/Origami-TinyML/tflm_hello_world/issues/45

**Comment by doyu** [comment]
This is related to https://github.com/Origami-TinyML/tflm_hello_world/issues/49#issuecomment-1436047019

**Comment by nellatuulikki** [comment]
Just want to clarify if this task is dependent on #93 being completed? If it is not dependent, am I understanding this task correctly that this could be implemented purely with streamlit?

**Comment by doyu** [comment]
I think that we go with the current implementation in Sprint 3 so that you could implement this independently basically. One point may be that, right now, Streamlit doesn't allow to store data locally. You may need to to use external storage for data (e.g. google driver or s3). For demo purpose, you could survive with running Streamlit locally using local storage, but for SaaS, it would run on a Cloud VM without local storage. You need to consider the design for that.

**Comment by nellatuulikki** [comment]
Are you thinking something like this https://docs.streamlit.io/knowledge-base/tutorials/databases/aws-s3 ?

**Comment by doyu** [comment]
> Are you thinking something like this https://docs.streamlit.io/knowledge-base/tutorials/databases/aws-s3 ?

Yes since Streamlit doesn't have any access to the local storage at all, at least, for now.

**Comment by nellatuulikki** [comment]
Ok, I'll try to set up aws s3 storage by following those instructions

**Comment by nellatuulikki** [comment]
I have now set up S3 bucket and manage to upload a csv file from there to streamlit by following those instructions. Next step is to store image data from streamlit to S3. Just for a clarification, do you want me to create a feature that stores only one image? 

I was thinking implementing something like this:

1.  User uploads an image in Data page
2.  When photo has been uploaded successfully, user can press button "Store image data"
3.  WebApp informs if the storing has been successfull

**Comment by doyu** [comment]
@nellatuulikki

it's really good that you asked before implementing without assuming something here.

> do you want me to create a feature that stores only one image?

No, data should be included in a compressed archive file. The sequence would be:

1. The original data archive is extracted in S3 storage. If not extract, just leave directory structure as it is in S3.
2. User uploads an image in Data page.
3. When the image has been uploaded successfully, user can press button "Store image data"
4. A new image data should be stored in the correct path, assuming the parent path is 'lablel', for example.
5. Compress dir to an archive if possible.
6. WebApp informs if the storing has been successfull

You  may want to take a look at how external data stored and processed in ML from :
https://github.com/fastai/fastai/blob/master/nbs/04_data.external.ipynb

Let me know if there's some inconsistency above. Thanks!

**Comment by nellatuulikki** [comment]
![Image](https://user-images.githubusercontent.com/94007460/221284559-990f43fe-de92-4a07-b1b3-7135deb84ebc.png)

**Comment by doyu** [comment]
Looks so cool! Great work!!

BTW, how did you choose a image? from a path or name?

>  If this looks ok, I could maybe tomorrow start to add the labeled images to S3.

Yes, please. Please don't hesitate to ask any questions.

**Comment by doyu** [comment]
@nellatuulikki would it be possible to switch backend storage dynamically, S3 or local one, via some environmental variable? It's better to reduce the external dependency, esp. for CI and demo.

**Comment by nellatuulikki** [comment]
I have now created PR with possibility to store only to S3. But basically you are asking that user could select if the images are stored locally or to S3? And do you mean that the app would create a new directory locally (if it doesn't exist already) and then store files there?

I was also wondering should we allow storing of unlabeled images? That would mean that there is third directory where all unlabeled images would be stored.

**Comment by doyu** [comment]
> I was also wondering should we allow storing of unlabeled images?

Yes

**Comment by FexbYk23** [comment]
> What I meant here was the same API to store image but this API backend should be changed on-the-fly with an enromental variable, S3 or local.

Would [localstack](https://github.com/localstack/localstack) be suitable for this? I was able to run Nella's code with it.

**Comment by doyu** [comment]
> Would localstack be suitable for this? I was able to run Nella's code with it.

@FexbYk23 @nellatuulikki 
Sounds what we want.
Please evaluate correctly if this is overengineering for our current usage or not.

**Comment by doyu** [comment]
One of another trends, https://world.hey.com/dhh/we-stand-to-save-7m-over-five-years-from-our-cloud-exit-53996caa

**Comment by FexbYk23** [comment]
> Please evaluate correctly if this is overengineering for our current usage or not.
> 
> BTW: do you have any idea how `localstack` intercepts AWS API call locally? (e.g. local port mapping?)

It doesn't intercept anything. It's a server that you run locally and you connect to it instead of the normal s3 servers.

It replicates the whole s3 system making it convenient for testing your code, but it's also quite heavy since it involves running another server. A more lightweight solution would probably be enough for our use case.

As I mentioned, I already used it to test Nella's code and I made a [branch with it set up](https://github.com/FexbYk23/tflm_hello_world/tree/localstack). I can create a pull request if we want to use it, if not, you can still use it for testing.

**Comment by doyu** [comment]
@FexbYk23 Would it be possible to switch without ipynb code? Just with MACRO in yml?

**Comment by FexbYk23** [comment]
> @FexbYk23 Would it be possible to switch without ipynb code? Just with MACRO in yml?

What do you mean by "MACRO in yml"?

**Comment by doyu** [comment]
Environmental var?

**Comment by FexbYk23** [comment]
> Environmental var?

Doesn't seem to be possible because [boto3 has no variable for setting endpoint url](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#using-environment-variables).

**Comment by doyu** [comment]
I meant, the core concept of localstack is to switch dynamically between mock and was with some environmental var? With localstack, you can do?

**Comment by FexbYk23** [comment]
> I meant, the core concept of localstack is to switch dynamically between mock and aws with some environmental var? With localstack, you can do?

I haven't seen anything about localstack being able to relay requests to real aws. Thus switching between mock(localstack) and aws would have to be done by our python application.

**Comment by lifeofborna** [comment]
Hiroshi agreed to look into this
